File: frontend/src/api/worker.ts
// 后端工人 API（走 @/api/http 统一 axios 客户端）。

import { ApiError, api } from '@/api/http'
import type {
  Worker,
  WorkerCreatePayload,
  WorkerListResult,
  WorkerUpdatePayload,
} from '@/types/worker'

export interface ListWorkersParams {
  name_like?: string
  is_active?: boolean
  limit?: number
  offset?: number
}

function cleanParams<T extends object>(p: T): Record<string, unknown> {
  const out: Record<string, unknown> = {}
  for (const [k, v] of Object.entries(p)) {
    if (v === undefined || v === null || v === '') continue
    out[k] = v
  }
  return out
}

export async function listWorkers(
  params: ListWorkersParams = {},
): Promise<WorkerListResult> {
  const resp = await api.get<WorkerListResult>('/workers', {
    params: cleanParams(params),
  })
  return resp.data
}

export async function getWorker(id: string): Promise<Worker> {
  const resp = await api.get<Worker>(`/workers/${id}`)
  return resp.data
}

export async function createWorker(payload: WorkerCreatePayload): Promise<Worker> {
  const resp = await api.post<Worker>('/workers', payload)
  return resp.data
}

export async function updateWorker(
  id: string,
  payload: WorkerUpdatePayload,
): Promise<Worker> {
  const resp = await api.post<Worker>(`/workers/${id}/update`, payload)
  return resp.data
}

export async function deactivateWorker(id: string): Promise<Worker> {
  const resp = await api.post<Worker>(`/workers/${id}/deactivate`)
  return resp.data
}

export async function reactivateWorker(id: string): Promise<Worker> {
  const resp = await api.post<Worker>(`/workers/${id}/reactivate`)
  return resp.data
}

// ============ 工牌扫码定位 ============
//
// 旧实现：拉一次 GET /workers?is_active=true&limit=500 → 客户端 Array.find。
// 问题：(1) 整张工人表被无权用户拿走（信息泄露 / 越权）；
//       (2) SHELF_ACCOUNT 根本无权调 GET /workers（router 强制 MANAGER），原本就是 403 隐患；
//       (3) 500 条硬上限导致工人 >500 时扫描误报；
//       (4) 60s TTL 导致新增 / 停用延迟生效。
// 新实现：POST /workers/verify-badge 单点 query；权限 = require_auth()，
//       MANAGER 与 SHELF_ACCOUNT 都能调。后端在 service 层做 is_active 校验。

// 后端错误码：20201 = BIZ_WORKER_NOT_FOUND, 20202 = BIZ_WORKER_INACTIVE。
// 这两种是扫描时的"未识别"业务态，前端按 null 处理；其他错误原样抛出。
const WORKER_NOT_FOUND = 20201
const WORKER_INACTIVE = 20202

/**
 * 按工牌码精确匹配工人。
 * - 命中且在职 → 返回 Worker。
 * - 不存在 / 已停用 → 返回 null（不抛错，调用方按业务决定提示文案）。
 * - 网络 / 其他错误 → 原样抛 ApiError。
 */
export async function findWorkerByBadge(badgeCode: string): Promise<Worker | null> {
  const code = badgeCode.trim()
  if (!code) return null

  try {
    const resp = await api.post<Worker>('/workers/verify-badge', {
      badge_code: code,
    })
    return resp.data
  } catch (e) {
    if (e instanceof ApiError && (e.code === WORKER_NOT_FOUND || e.code === WORKER_INACTIVE)) {
      return null
    }
    throw e
  }
}

File: api/v1/worker.py
from fastapi import APIRouter, Depends, Query, status as http_status

from api.deps import get_worker_service
from core.permission import require_role
from model.enums import UserRole
from schema.worker import (
    WorkerCreateRequest,
    WorkerListOut,
    WorkerListQuery,
    WorkerOut,
    WorkerUpdateRequest,
)
from service import WorkerService

router = APIRouter(
    prefix="/workers",
    tags=["工人管理"],
    dependencies=[Depends(require_role(UserRole.MANAGER))],
)


@router.get(
    "",
    response_model=WorkerListOut,
    summary="工人列表",
)
async def list_workers(
    name_like: str | None = Query(default=None, description="姓名模糊匹配"),
    is_active: bool | None = Query(default=None, description="是否在职"),
    limit: int = Query(default=50, ge=1, le=500),
    offset: int = Query(default=0, ge=0),
    svc: WorkerService = Depends(get_worker_service),
) -> WorkerListOut:
    return await svc.list_workers(
        WorkerListQuery(
            name_like=name_like,
            is_active=is_active,
            limit=limit,
            offset=offset,
        )
    )


@router.post(
    "",
    response_model=WorkerOut,
    status_code=http_status.HTTP_201_CREATED,
    summary="新增工人（录入工牌码 + 姓名）",
)
async def create_worker(
    payload: WorkerCreateRequest,
    svc: WorkerService = Depends(get_worker_service),
) -> WorkerOut:
    return await svc.create_worker(payload)


@router.get(
    "/{worker_id}",
    response_model=WorkerOut,
    summary="工人详情",
)
async def get_worker(
    worker_id: int,
    svc: WorkerService = Depends(get_worker_service),
) -> WorkerOut:
    return await svc.get_worker(worker_id)


@router.post(
    "/{worker_id}/update",
    response_model=WorkerOut,
    summary="更新工人字段（仅传需要改的）",
)
async def update_worker(
    worker_id: int,
    payload: WorkerUpdateRequest,
    svc: WorkerService = Depends(get_worker_service),
) -> WorkerOut:
    return await svc.update_worker(worker_id, payload)


@router.post(
    "/{worker_id}/deactivate",
    response_model=WorkerOut,
    summary="停用工人（软删 + is_active=false）",
)
async def deactivate_worker(
    worker_id: int,
    svc: WorkerService = Depends(get_worker_service),
) -> WorkerOut:
    return await svc.deactivate(worker_id)


@router.post(
    "/{worker_id}/reactivate",
    response_model=WorkerOut,
    summary="重新启用工人",
)
async def reactivate_worker(
    worker_id: int,
    svc: WorkerService = Depends(get_worker_service),
) -> WorkerOut:
    return await svc.reactivate(worker_id)